File: 2021_22 - Practicum in Object Oriented Programming/Olympic Games Statistics App/CLI App/src/Medal.cpp
#include "Medal.hpp"

#include <string>  // std::strcmp


// ================================
// Constructors and Destructor:


sale::Medal::Medal(const TMedal& medal, const int olympic_games_id,
				   std::shared_ptr<const std::string> country_name,
				   std::shared_ptr<sale::Event> event_)
	: medal(medal), olympic_games_id(olympic_games_id),
	  country_name(std::move(country_name)), event_(std::move(event_))
{
}

sale::Medal::Medal(const std::string& medal_str, const int olympic_games_id,
				   std::shared_ptr<const std::string> country_name,
				   std::shared_ptr<sale::Event> event_)
	: Medal(medal_str.c_str(), olympic_games_id,
			std::move(country_name), std::move(event_))
{
}

sale::Medal::Medal(const char* const medal_str, const int olympic_games_id,
				   std::shared_ptr<const std::string> country_name,
				   std::shared_ptr<sale::Event> event_)
	: olympic_games_id(olympic_games_id), country_name(std::move(country_name)),
	  event_(std::move(event_))
{
	if (std::strcmp(medal_str, "") == 0) {
		this->medal = sale::Medal::TMedal::NoMedal;
	}
	else if (std::strcmp(medal_str, "Gold") == 0) {
		this->medal = sale::Medal::TMedal::Gold;
	}
	else if (std::strcmp(medal_str, "Silver") == 0) {
		this->medal = sale::Medal::TMedal::Silver;
	}
	else if (std::strcmp(medal_str, "Bronze") == 0) {
		this->medal = sale::Medal::TMedal::Bronze;
	}
	else if (std::strcmp(medal_str, "No Medal") == 0) {
		this->medal = sale::Medal::TMedal::NoMedal;
	}
	else {
		throw InvalidMedalStrFormat();
	}
}


// ================================
// Operator Overloads:


bool sale::Medal::operator==(const Medal& other) const
{
	return this->medal == other.medal;
}

bool sale::Medal::operator!=(const Medal& other) const
{
	return !(*this == other);
}

bool sale::Medal::operator>(const Medal& other) const
{
	return this->medal < other.medal;
}

bool sale::Medal::operator<(const Medal& other) const
{
	return this->medal > other.medal;
}

bool sale::Medal::operator>=(const Medal& other) const
{
	return !(*this < other);
}

bool sale::Medal::operator<=(const Medal& other) const
{
	return !(*this > other);
}

int sale::Medal::getMedalScore() const
{
	return (int) sale::Medal::TMedal::NoMedal - (int) this->medal;
}

int sale::Medal::getOlympicGamesID() const
{
	return this->olympic_games_id;
}

sale::Medal::TMedal sale::Medal::getMedal() const
{
	return this->medal;
}

bool sale::Medal::isSharedByTeam() const
{
	return this->is_shared_by_team;
}

const std::string& sale::Medal::getCountryNameReference() const
{
	return *this->country_name;
}

std::shared_ptr<const std::string> sale::Medal::getCountryName() const
{
	return this->country_name;
}

std::shared_ptr<sale::Event> sale::Medal::getEvent() const
{
	return this->event_;
}

void sale::Medal::setAsSharedByTeam()
{
	this->is_shared_by_team = true;
}

File: 2021_22 - Practicum in Object Oriented Programming/Olympic Games Statistics App/GUI App/Java/src/awards/Medal.java
package awards;

import java.util.Objects;
import exceptions.InvalidMedalStrFormat;
import olympics.events.Event;


public class Medal implements Comparable<Medal> {

	public enum TMedal {
		Gold(0), Silver(1), Bronze(2), NoMedal(3);

		private int value;
		private TMedal(final int value) { this.value = value; }
		private int getValue() { return this.value; }
	};


	// ==========================================
	// Attributes:


	private TMedal medal = Medal.TMedal.NoMedal;
	private int olympic_games_id;
	private boolean is_shared_by_team = false;
	private String country_name;
	private Event event_;


	// ==========================================
    // Constructors:

	
	public Medal(TMedal medal, final int oly_games_id,
				String country_name, Event event_)
	{
		this.medal = medal;
		this.olympic_games_id = oly_games_id;
		this.country_name = country_name;
		this.event_ = event_;
	}


	public Medal(String medal_str, final int oly_games_id,
				 String country_name, Event event_)
		throws InvalidMedalStrFormat
	{
		this.olympic_games_id = oly_games_id;
		this.country_name = country_name;
		this.event_ = event_;

		if (medal_str.equals("")) {
			this.medal = TMedal.NoMedal;
		} else if (medal_str.equals("Gold")) {
			this.medal = TMedal.Gold;
		} else if (medal_str.equals("Silver")) {
			this.medal = TMedal.Silver;
		} else if (medal_str.equals("Bronze")) {
			this.medal = TMedal.Bronze;
		} else {
			throw new InvalidMedalStrFormat();
		}
	}


	// ==========================================
	// Getters:


	public int getMedalScore() {
		return this.medal.getValue();
	}

	public int getOlympicGamesID() {
		return this.olympic_games_id;
	}

	public TMedal getMedal() {
		return this.medal;
	}

	public boolean isSharedByTeam() {
		return this.is_shared_by_team;
	}

	public String getCountryName() {
		return this.country_name;
	}
	
	public Event getEvent() {
		return this.event_;
	}


	// ==========================================
	// Setters:


	public void setAsSharedByTeam() {
		this.is_shared_by_team = true;
	}

	
	// ==========================================
	// Comparisons:


	@Override
	public int hashCode() {
		return Objects.hash(medal);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		Medal other = (Medal) obj;
		return medal.getValue() == other.medal.getValue();
	}

	@Override
	public int compareTo(Medal other) {
		return other.medal.getValue() - this.medal.getValue();
	}
	
	@Override
	public String toString() {
		return this.medal.toString();
	}

}
